# Operations Challenge - Kishan Chandrakumar
_June 2020_

## Task 1 - Improving KYC

Revolut is legally bound to ensure KYC checks are adhered to. However, to provide a seamless experience for customers, we explore why many customers may have been prevented from passing the automatic checks before opening an account. 

In [1]:
import pandas as pd #Faster processing ability than using for loops
print(pd.__version__)

1.0.1


In [2]:
faces = pd.read_csv('facial_similarity_reports.csv')
docs = pd.read_csv('doc_reports.csv')
print('Facial Similarity data has dimensions: ', faces.shape) 
print('Documents data has dimensions: ', docs.shape) 
#Both datasets have the same number of rows, suggesting same customers

Facial Similarity data has dimensions:  (176404, 9)
Documents data has dimensions:  (176404, 19)


In [3]:
print('\n Facial Similarity data info: ')
print(faces.info())
print('\n Documents data has info: ')
print(docs.info())


 Facial Similarity data info: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 176404 entries, 0 to 176403
Data columns (total 9 columns):
 #   Column                         Non-Null Count   Dtype 
---  ------                         --------------   ----- 
 0   Unnamed: 0                     176404 non-null  int64 
 1   user_id                        176404 non-null  object
 2   result                         176403 non-null  object
 3   face_comparison_result         166007 non-null  object
 4   created_at                     176404 non-null  object
 5   facial_image_integrity_result  175941 non-null  object
 6   visual_authenticity_result     150290 non-null  object
 7   properties                     176404 non-null  object
 8   attempt_id                     176404 non-null  object
dtypes: int64(1), object(8)
memory usage: 12.1+ MB
None

 Documents data has info: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 176404 entries, 0 to 176403
Data columns (total 19 columns):
 # 

We first look at some summary data of both datasets to identify problems and understand the data more. We can already see that some 'user_id' values are repeated, with the most common user having 6 rows in total. 

The first row of facial data shows something unusual. 'visual_authenticity_result' returns 'consider', however the 'result' column still reads 'clear'. This may highlight customers 'passing' when they are not supposed to, which opens up a potential for regulatory penalties, although this is wider than the scope of this task.

In [4]:
faces.describe(include = 'all') 


Unnamed: 0                           user_id  result  \
count   176404.000000                            176404  176403   
unique            NaN                            142724       2   
top               NaN  6eeb7dbdf1fa4e7c95413bc0608dd21c   clear   
freq              NaN                                 6  165486   
mean     91515.317476                               NaN     NaN   
std      52664.122568                               NaN     NaN   
min          0.000000                               NaN     NaN   
25%      45660.750000                               NaN     NaN   
50%      91487.500000                               NaN     NaN   
75%     137570.500000                               NaN     NaN   
max     181991.000000                               NaN     NaN   

       face_comparison_result            created_at  \
count                  166007                176404   
unique                      2                169820   
top                     clear  2017-07-27T10:09:03Z   
freq                   165485                     6   
mean                      NaN                   NaN   
std                       NaN                   NaN   
min                       NaN                   NaN   
25%                       NaN                   NaN   
50%                       NaN                   NaN   
75%                       NaN                   NaN   
max                       NaN                   NaN   

       facial_image_integrity_result visual_authenticity_result properties  \
count                         175941                     150290     176404   
unique                             2                          2         80   
top                            clear                      clear         {}   
freq                          166004                     147260     172921   
mean                             NaN                        NaN        NaN   
std                              NaN                        NaN        NaN   
min                              NaN                        NaN        NaN   
25%                              NaN                        NaN        NaN   
50%                              NaN                        NaN        NaN   
75%                              NaN                        NaN        NaN   
max                              NaN                        NaN        NaN   

                              attempt_id  
count                             176404  
unique                            176404  
top     900ff035bbde4a78970ce661fb0776f6  
freq                                   1  
mean                                 NaN  
std                                  NaN  
min                                  NaN  
25%                                  NaN  
50%                                  NaN  
75%                                  NaN  
max                                  NaN

In [5]:
docs.describe(include = 'all') 

Unnamed: 0                           user_id  result  \
count   176404.000000                            176404  176404   
unique            NaN                            142724       2   
top               NaN  6eeb7dbdf1fa4e7c95413bc0608dd21c   clear   
freq              NaN                                 6  132402   
mean     91515.317476                               NaN     NaN   
std      52664.122568                               NaN     NaN   
min          0.000000                               NaN     NaN   
25%      45660.750000                               NaN     NaN   
50%      91487.500000                               NaN     NaN   
75%     137570.500000                               NaN     NaN   
max     181991.000000                               NaN     NaN   

       visual_authenticity_result image_integrity_result  \
count                      150290                 176403   
unique                          2                      2   
top                         clear                  clear   
freq                       147260                 136660   
mean                          NaN                    NaN   
std                           NaN                    NaN   
min                           NaN                    NaN   
25%                           NaN                    NaN   
50%                           NaN                    NaN   
75%                           NaN                    NaN   
max                           NaN                    NaN   

       face_detection_result image_quality_result            created_at  \
count                 150261               176403                176404   
unique                     2                    2                169810   
top                    clear                clear  2017-07-27T10:09:03Z   
freq                  149687               152000                     6   
mean                     NaN                  NaN                   NaN   
std                      NaN                  NaN                   NaN   
min                      NaN                  NaN                   NaN   
25%                      NaN                  NaN                   NaN   
50%                      NaN                  NaN                   NaN   
75%                      NaN                  NaN                   NaN   
max                      NaN                  NaN                   NaN   

       supported_document_result conclusive_document_quality_result  \
count                     175900                              95217   
unique                         2                                  2   
top                        clear                              clear   
freq                      174213                              81652   
mean                         NaN                                NaN   
std                          NaN                                NaN   
min                          NaN                                NaN   
25%                          NaN                                NaN   
50%                          NaN                                NaN   
75%                          NaN                                NaN   
max                          NaN                                NaN   

       colour_picture_result data_validation_result data_consistency_result  \
count                  95222                 142974                   92229   
unique                     2                      2                       2   
top                    clear                  clear                   clear   
freq                   95118                 141335                   92059   
mean                     NaN                    NaN                     NaN   
std                      NaN                    NaN                     NaN   
min                      NaN                    NaN                     NaN   
25%                      NaN                    NaN                     NaN   
50%                 

In [6]:
faces.head()

Unnamed: 0                           user_id result face_comparison_result  \
0           0  ab23fae164e34af0a1ad1423ce9fd9f0  clear                  clear   
1           1  15a84e8951254011b47412fa4e8f65b8  clear                  clear   
2           2  ffb82fda52b041e4b9af9cb4ef298c85  clear                  clear   
3           3  bd4a8b3e3601427e88aa1d9eab9f4290  clear                  clear   
4           4  f52ad1c7e69543a9940c3e7f8ed28a39  clear                  clear   

             created_at facial_image_integrity_result  \
0  2017-06-20T23:12:58Z                         clear   
1  2017-06-20T23:16:04Z                         clear   
2  2017-06-20T17:59:49Z                         clear   
3  2017-06-20T17:59:39Z                         clear   
4  2017-06-20T18:08:09Z                         clear   

  visual_authenticity_result properties                        attempt_id  
0                   consider         {}  050a0596de424fab83c433eaa18b3f8d  
1                      clear         {}  f69c1e5f45a64e50a26740b9bfb978b7  
2                      clear         {}  f9f84f3055714d8e8f7419dc984d1769  
3                      clear         {}  10a54a1ecf794404be959e030f11fef6  
4                      clear         {}  1f320d1d07de493292b7e0d5ebfb1cb9

In [7]:
docs.head()

Unnamed: 0                           user_id    result  \
0           0  ab23fae164e34af0a1ad1423ce9fd9f0  consider   
1           1  15a84e8951254011b47412fa4e8f65b8     clear   
2           2  ffb82fda52b041e4b9af9cb4ef298c85     clear   
3           3  bd4a8b3e3601427e88aa1d9eab9f4290     clear   
4           4  f52ad1c7e69543a9940c3e7f8ed28a39     clear   

  visual_authenticity_result image_integrity_result face_detection_result  \
0                   consider                  clear                 clear   
1                      clear                  clear                 clear   
2                      clear                  clear                 clear   
3                      clear                  clear                 clear   
4                      clear                  clear                 clear   

  image_quality_result            created_at supported_document_result  \
0                clear  2017-06-20T23:12:57Z                     clear   
1                clear  2017-06-20T23:16:04Z                     clear   
2                clear  2017-06-20T17:59:49Z                     clear   
3                clear  2017-06-20T17:59:38Z                     clear   
4                clear  2017-06-20T18:08:09Z                     clear   

  conclusive_document_quality_result colour_picture_result  \
0                                NaN                   NaN   
1                                NaN                   NaN   
2                                NaN                   NaN   
3                                NaN                   NaN   
4                                NaN                   NaN   

  data_validation_result data_consistency_result data_comparison_result  \
0                  clear                   clear                    NaN   
1                  clear                     NaN                    NaN   
2                  clear                   clear                    NaN   
3                  clear                   clear                    NaN   
4                  clear                   clear                    NaN   

                         attempt_id police_record_result  \
0  050a0596de424fab83c433eaa18b3f8d                clear   
1  f69c1e5f45a64e50a26740b9bfb978b7                clear   
2  f9f84f3055714d8e8f7419dc984d1769                clear   
3  10a54a1ecf794404be959e030f11fef6                clear   
4  1f320d1d07de493292b7e0d5ebfb1cb9                clear   

  compromised_document_result  \
0                         NaN   
1                         NaN   
2                         NaN   
3                         NaN   
4                         NaN   

                                          properties sub_result  
0  {'gender': 'Male', 'nationality': 'IRL', 'docu...    caution  
1  {'gender': 'Female', 'document_type': 'driving...      clear  
2  {'gender': 'Male', 'nationality': 'ITA', 'docu...      clear  
3  {'gender': 'Male', 'issuing_date': '2007-08', ...      clear  
4  {'gender': 'Male', 'nationality': 'POL', 'docu...      clear

If a customer is clear for all elements of both verification stages, then the customer 'passes'. Therefore we could combine the data to identify which columns have the most bottlenecks.

#### Merging both tables

In [8]:
faces['user_id'].equals(docs['user_id']) #both datasets correspond to same customer base so we can merge

True

In [9]:
trial = faces.merge(docs, on=faces.index,)
trial.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 176404 entries, 0 to 176403
Data columns (total 29 columns):
 #   Column                              Non-Null Count   Dtype 
---  ------                              --------------   ----- 
 0   key_0                               176404 non-null  int64 
 1   Unnamed: 0_x                        176404 non-null  int64 
 2   user_id_x                           176404 non-null  object
 3   result_x                            176403 non-null  object
 4   face_comparison_result              166007 non-null  object
 5   created_at_x                        176404 non-null  object
 6   facial_image_integrity_result       175941 non-null  object
 7   visual_authenticity_result_x        150290 non-null  object
 8   properties_x                        176404 non-null  object
 9   attempt_id_x                        176404 non-null  object
 10  Unnamed: 0_y                        176404 non-null  int64 
 11  user_id_y                           176

Some columns can safely be dropped to make working with our data easier.

In [17]:
merged = trial.drop(["key_0", "Unnamed: 0_x", "Unnamed: 0_y", "user_id_y"], axis = 'columns')
merged.shape

(176404, 25)

In [18]:
merged = merged.drop_duplicates(subset = "user_id_x")

In [19]:
merged.shape

(142724, 25)

Now we want to analyse the failure data. All the 'Clear' ones can be put to one side, and we explore the 'Consider' rows by isolating them. 'result_x' and 'result_y' are most important, as either of those reading 'consider' will prevent a customer passing.

In [20]:
merged_consider = merged[(merged.result_x != 'clear') | (merged.result_y != 'clear')]
merged_consider.shape # we have 29645 rows that would not have 'passed'

(29645, 25)

In [23]:
merged_consider.describe(include='all')

user_id_x result_x face_comparison_result  \
count                              29645    29644                  23208   
unique                             29645        2                      2   
top     5964524d50ce4c2dbebdcbfb65cd9848    clear                  clear   
freq                                   1    22845                  22844   

                created_at_x facial_image_integrity_result  \
count                  29645                         29367   
unique                 29554                             2   
top     2017-10-11T17:06:48Z                         clear   
freq                       2                         23206   

       visual_authenticity_result_x properties_x  \
count                         14357        29645   
unique                            2           74   
top                           clear           {}   
freq                          12564        29317   

                            attempt_id_x  result_y  \
count                              29645     29645   
unique                             29645         2   
top     81e61dd1167b4ab6ac91e9ef5c585089  consider   
freq                                   1     26073   

       visual_authenticity_result_y  ... conclusive_document_quality_result  \
count                         14357  ...                              11154   
unique                            2  ...                                  2   
top                           clear  ...                           consider   
freq                          12564  ...                               8154   

       colour_picture_result data_validation_result data_consistency_result  \
count                  11154                  13649                    8530   
unique                     2                      2                       2   
top                    clear                  clear                   clear   
freq                   11089                  12660                    8417   

       data_comparison_result                      attempt_id_y  \
count                     423                             29645   
unique                      2                             29645   
top                     clear  81e61dd1167b4ab6ac91e9ef5c585089   
freq                      298                                 1   

       police_record_result compromised_document_result properties_y  \
count                 13852                        7172        29645   
unique                    2                           2        12766   
top                   clear                       clear           {}   
freq                  13837                        7160        15279   

       sub_result  
count       29645  
unique          4  
top      rejected  
freq        15278  

[4 rows x 25 columns]

We can immediately see that the document checks is the bottleneck, with 'result_y' having more 'consider' values than 'clear'. In particular, we can see from the following descriptions that 'image_integrity_result' is our top candidate for the failure, with 'consider' showing 23,489 times. 'conclusive_document_quality_result' comes in second with 8154 instances. 

The documentation states that both of these factors refer to whether the quality of the image is sufficient to ascertain whether the document is fraudulent or not. 

Revolut could work with Veritas and Revolut App engineers on:
* automatic performance check of the customer camera quality 
* offer guidance on how to avoid shadow and/or glare for first time pass rates for clients

In [26]:
merged_consider.loc[:,"visual_authenticity_result_y":"image_quality_result"].describe()

visual_authenticity_result_y image_integrity_result  \
count                         14357                  29645   
unique                            2                      2   
top                           clear               consider   
freq                          12564                  23489   

       face_detection_result image_quality_result  
count                  14353                29645  
unique                     2                    2  
top                    clear                clear  
freq                   14011                15365

In [27]:
merged_consider.loc[:,"supported_document_result":"data_comparison_result"].describe()

supported_document_result conclusive_document_quality_result  \
count                      29322                              11154   
unique                         2                                  2   
top                        clear                           consider   
freq                       28324                               8154   

       colour_picture_result data_validation_result data_consistency_result  \
count                  11154                  13649                    8530   
unique                     2                      2                       2   
top                    clear                  clear                   clear   
freq                   11089                  12660                    8417   

       data_comparison_result  
count                     423  
unique                      2  
top                     clear  
freq                      298

In [28]:
merged_consider.loc[:,"police_record_result":"compromised_document_result"].describe()

police_record_result compromised_document_result
count                 13852                        7172
unique                    2                           2
top                   clear                       clear
freq                  13837                        7160